bson/src/test/unit/org/bson/codecs/CollectionCodecSpecification.groovy

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.codecs
+
+import org.bson.BsonArray
+import org.bson.BsonDateTime
+import org.bson.BsonDocument
+import org.bson.BsonDocumentReader
+import org.bson.BsonDocumentWriter
+import org.bson.codecs.jsr310.Jsr310CodecProvider
+import org.bson.types.Binary
+import spock.lang.Specification
+import spock.lang.Unroll
+
+import java.lang.reflect.ParameterizedType
+import java.time.Instant
+
+import static org.bson.BsonDocument.parse
+import static org.bson.UuidRepresentation.C_SHARP_LEGACY
+import static org.bson.UuidRepresentation.JAVA_LEGACY
+import static org.bson.UuidRepresentation.PYTHON_LEGACY
+import static org.bson.UuidRepresentation.STANDARD
+import static org.bson.UuidRepresentation.UNSPECIFIED
+import static org.bson.codecs.configuration.CodecRegistries.fromCodecs
+import static org.bson.codecs.configuration.CodecRegistries.fromProviders
+import static org.bson.codecs.configuration.CodecRegistries.fromRegistries
+
+class CollectionCodecSpecification extends Specification {
```

Comment:
Currently these tests pass if the `else` branch in `AbstractCollectionCodec` is replaced with `throw Assertions.fail();`. It appears that we lack a test that decodes to a custom `Collection` that has a no-arg constructor. `CopyOnWriteArrayList` can be used as such a custom `Collection` in the `should decode to specified generic class` test.

Note that a similar thing is **not** applicable to `MapCodecV2Specification`, as it tests against `WeakHashMap`.

Reply:
Done: added CopyOnWriteArrayList to the where clause of the test.  The only lack of coverage now is the exceptional conditions.  Not sure it's worth it to cover those though.

Reply:
🤷‍♂️ I am fine with not having those